addons/web/static/tests/views/fields/reference_field_tests.js

```diff
@@ -598,6 +598,45 @@ QUnit.module("Fields", (hooks) => {
         await clickSave(target);
     });
 
+    QUnit.debug("ReferenceField with model_field option2", async function (assert) {
+        serverData.models.partner.onchanges = {
+            color(obj) {
+                if (obj.color === "black") {
+                    obj.model_id = 20;
+                    obj.reference = "product,37";
+                } else {
+                    obj.model_id = 17;
+                    obj.reference = "partner,1";
+                }
+            },
+        };
+
+        await makeView({
+            type: "form",
+            resModel: "partner",
+            resId: 1,
+            serverData,
+            arch: `
+                <form>
+                    <field name="color" />
+                    <field name="model_id" invisible="1"/>
+                    <field name="reference" options="{'model_field': 'model_id'}" />
+                </form>`,
+            mockRPC(route, { args, method }) {
+                if (method === "write") {
+                    assert.strictEqual(args[1].reference, "partner,4");
```

Comment:
For a sanity check, ensure that this route is indeed called. Either with verifySteps or with an expect